solr/core/src/java/org/apache/solr/request/SolrRequestInfo.java

```diff
@@ -52,35 +54,44 @@
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
 
   public static SolrRequestInfo getRequestInfo() {
-    return threadLocal.get();
+    if (threadLocal.get().isEmpty()) return null;
+    return threadLocal.get().peek();
   }
 
   public static void setRequestInfo(SolrRequestInfo info) {
-    // TODO: temporary sanity check... this can be changed to just an assert in the future
-    SolrRequestInfo prev = threadLocal.get();
-    if (prev != null) {
-      log.error("Previous SolrRequestInfo was not closed!  req={}", prev.req.getOriginalParams());
-      log.error("prev == info : {}", prev.req == info.req, new RuntimeException());
+    if (info == null) {
+      throw new IllegalArgumentException("SolrRequestInfo is null");
+    } else {
+      threadLocal.get().push(info);
```

Comment:
Shouldn't we limit maximum stack depth?
What if one component leak SolrRequestInfo accidentally, like push it on every request but refuse to pop it when  exception occur. Thus it may steadily flood heap. Or I miss something?

Reply:
Good idea!

Reply:
what would you suggest for the stack size limit?

Reply:
100 for sure